// cig-data-patcher/dist/utils/mappers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mfaAuthenticationInfoMapper = exports.privilegedMapper = exports.deviceMapper = exports.sessionMapper = void 0;
const sessionMapper = ({ session_name, session_id }) => ({
    cookie: `${session_name}`,
    key: `X-${session_name}`,
    value: session_id,
});
exports.sessionMapper = sessionMapper;
const deviceMapper = ({ device_header, device_id }) => ({
    cookie: '_rsi_device',
    key: `${device_header}`,
    value: device_id,
});
exports.deviceMapper = deviceMapper;
const privilegedMapper = ({ badges }) => {
    if (badges) {
        const names = Object.values(badges);
        return names.indexOf('Staff') >= 0 || names.indexOf('QA') >= 0;
    }
    return false;
};
exports.privilegedMapper = privilegedMapper;
const mfaAuthenticationInfoMapper = ({ device_id, new_device, prompt, device_name, device_type, }) => {
    const registered = !device_id && !new_device;
    const type = prompt.includes('authenticator') ? 'authenticator' : 'email';
    return {
        type,
        deviceInfo: {
            registered,
            name: registered ? device_name : 'My Device',
            type: device_type,
        },
    };
};
exports.mfaAuthenticationInfoMapper = mfaAuthenticationInfoMapper;
