// node-multi-downloader/dist/mdFile/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyDownloadedChunk = exports.download = exports.close = exports.open = exports.create = exports.createMDFileVerificationHeader = void 0;
__exportStar(require("./constants"), exports);
var createVerificationHeader_1 = require("./createVerificationHeader");
Object.defineProperty(exports, "createMDFileVerificationHeader", { enumerable: true, get: function () { return createVerificationHeader_1.createMDFileVerificationHeader; } });
var create_1 = require("./create");
Object.defineProperty(exports, "create", { enumerable: true, get: function () { return create_1.create; } });
var open_1 = require("./open");
Object.defineProperty(exports, "open", { enumerable: true, get: function () { return open_1.open; } });
var close_1 = require("./close");
Object.defineProperty(exports, "close", { enumerable: true, get: function () { return close_1.close; } });
var download_1 = require("./download");
Object.defineProperty(exports, "download", { enumerable: true, get: function () { return download_1.download; } });
Object.defineProperty(exports, "verifyDownloadedChunk", { enumerable: true, get: function () { return download_1.verifyDownloadedChunk; } });
