// node-multi-downloader/dist/cli/utils/downloadPhase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PHASE = void 0;
exports.PHASE = {
    INITIAL_DOWNLOAD_START: 'installer@initial-download-start',
    INITIAL_DOWNLOAD_PROGRESS: 'installer@initial-download-progress',
    INITIAL_DOWNLOAD_END: 'installer@initial-download-end',
    RETRIEVE_LOCAL_FILE_LIST_START: 'installer@retrieve-local-file-list-start',
    RETRIEVE_LOCAL_FILE_LIST_END: 'installer@retrieve-local-file-list-end',
    RETRIEVE_REMOTE_FILE_LIST_START: 'installer@retrieve-remote-file-list-start',
    RETRIEVE_REMOTE_FILE_LIST_PROGRESS: 'installer@retrieve-remote-file-list-progress',
    RETRIEVE_REMOTE_FILE_LIST_END: 'installer@retrieve-remote-file-list-end',
    COMPUTE_FILE_LIST_DIFFERENCE_START: 'installer@compute-file-list-difference-start',
    COMPUTE_FILE_LIST_DIFFERENCE_END: 'installer@compute-file-list-difference-end',
    UPDATE_LOOSE_FILES_START: 'installer@update-loose-files-start',
    UPDATE_LOOSE_FILES_PROGRESS: 'installer@update-loose-files-progress',
    UPDATE_LOOSE_FILES_END: 'installer@update-loose-files-end',
    UPDATE_P4K_STRUCTURE_START: 'installer@update-p4k-structure-start',
    UPDATE_P4K_STRUCTURE_END: 'installer@update-p4k-structure-end',
    UPDATE_FILES_INSIDE_P4K_START: 'installer@update-files-inside-p4k-start',
    UPDATE_FILES_INSIDE_P4K_PROGRESS: 'installer@update-files-inside-p4k-progress',
    UPDATE_FILES_INSIDE_P4K_END: 'installer@update-files-inside-p4k-end',
};
