// cig-data-patcher/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstallMode = exports.InstallerState = exports.InstallerPhaseEvents = exports.InstallerOptions = void 0;
var InstallerOptions;
(function (InstallerOptions) {
    InstallerOptions["ConcurrentTransfers"] = "concurrentTransfers";
    InstallerOptions["MaximumDownloadBandwidth"] = "maximumDownloadBandwidth";
    InstallerOptions["MaximumTransfersInFlight"] = "maximumTransfersInFlight";
    InstallerOptions["RetriesPerFile"] = "retriesPerFile";
})(InstallerOptions || (exports.InstallerOptions = InstallerOptions = {}));
var InstallerPhaseEvents;
(function (InstallerPhaseEvents) {
    InstallerPhaseEvents["OpenP4kStart"] = "installer@open-p4k-start";
    InstallerPhaseEvents["OpenP4kEnd"] = "installer@open-p4k-end";
    InstallerPhaseEvents["RetrieveLocalFileListStart"] = "installer@retrieve-local-file-list-start";
    InstallerPhaseEvents["RetrieveLocalFileListEnd"] = "installer@retrieve-local-file-list-end";
    InstallerPhaseEvents["RetrieveRemoteFileListStart"] = "installer@retrieve-remote-file-list-start";
    InstallerPhaseEvents["RetrieveRemoteFileListProgress"] = "installer@retrieve-remote-file-list-progress";
    InstallerPhaseEvents["RetrieveRemoteFileListEnd"] = "installer@retrieve-remote-file-list-end";
    InstallerPhaseEvents["ComputeFileListDifferenceStart"] = "installer@compute-file-list-difference-start";
    InstallerPhaseEvents["ComputeFileListDifferenceEnd"] = "installer@compute-file-list-difference-end";
    InstallerPhaseEvents["UpdateLooseFilesStart"] = "installer@update-loose-files-start";
    InstallerPhaseEvents["UpdateLooseFilesProgress"] = "installer@update-loose-files-progress";
    InstallerPhaseEvents["UpdateLooseFilesEnd"] = "installer@update-loose-files-end";
    InstallerPhaseEvents["UpdateP4kStructureStart"] = "installer@update-p4k-structure-start";
    InstallerPhaseEvents["UpdateP4kStructureEnd"] = "installer@update-p4k-structure-end";
    InstallerPhaseEvents["UpdateFilesInsideP4kStart"] = "installer@update-files-inside-p4k-start";
    InstallerPhaseEvents["UpdateFilesInsideP4kProgress"] = "installer@update-files-inside-p4k-progress";
    InstallerPhaseEvents["UpdateFilesInsideP4kEnd"] = "installer@update-files-inside-p4k-end";
    InstallerPhaseEvents["P4kVerificationStart"] = "installer@p4k-verification-start";
    InstallerPhaseEvents["P4kVerificationProgress"] = "installer@p4k-verification-progress";
    InstallerPhaseEvents["P4kVerificationEnd"] = "installer@p4k-verification-end";
    InstallerPhaseEvents["LooseFileVerificationStart"] = "installer@loose-file-verification-start";
    InstallerPhaseEvents["LooseFileVerificationProgress"] = "installer@loose-file-verification-progress";
    InstallerPhaseEvents["LooseFileVerificationEnd"] = "installer@loose-file-verification-end";
})(InstallerPhaseEvents || (exports.InstallerPhaseEvents = InstallerPhaseEvents = {}));
var InstallerState;
(function (InstallerState) {
    InstallerState["Idle"] = "idle";
    InstallerState["Installing"] = "installing";
    InstallerState["Verifying"] = "verifying";
    InstallerState["Paused"] = "paused";
    InstallerState["Canceled"] = "canceled";
    InstallerState["Error"] = "error";
})(InstallerState || (exports.InstallerState = InstallerState = {}));
var InstallMode;
(function (InstallMode) {
    InstallMode["Install"] = "install";
    InstallMode["Verify"] = "verify";
})(InstallMode || (exports.InstallMode = InstallMode = {}));
