// cig-data-patcher/dist/api/games.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class RSIApiGames {
    constructor(api) {
        this.libraryTimeout = 10000;
        this.api = api;
    }
    async getClaims() {
        const response = await this.api.call({
            endpoint: this.api.endpoints.GAMES_CLAIMS,
        });
        return response.data;
    }
    async getLibrary(claims) {
        const response = await this.api.call({
            endpoint: this.api.endpoints.GAMES_LIBRARY,
            payload: {
                claims,
            },
        });
        const games = response.data.games.map((game) => {
            const { id: gameId, name: gameName, channels: gameChannels } = game;
            const channels = Object.values(gameChannels).map(({ id, name, version, versionLabel, platformId, servicesEndpoint, nid, network }) => ({
                id,
                name,
                version,
                versionLabel,
                platformId,
                servicesEndpoint,
                nid,
                network,
            }));
            return { id: gameId, name: gameName, channels };
        });
        return games;
    }
    async getRelease({ claims, gameID, channelID, platformID, }) {
        const response = await this.api.call({
            endpoint: this.api.endpoints.GAMES_RELEASE,
            payload: {
                claims,
                gameId: gameID,
                channelId: channelID,
                platformId: platformID,
            },
        });
        return response.data;
    }
}
exports.default = RSIApiGames;
