// @rsilauncher/app-shared/lib/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const configuration = {
    env: {
        local: {
            apiUrl: 'https://rsi-platform.local.dev/api/launcher/v3',
            rootUrl: 'https://rsi-platform.local.dev/heap/',
        },
        qa1: {
            apiUrl: 'https://qa-1.cloudimperiumgames.com/api/launcher/v3',
            rootUrl: 'https://qa-1.cloudimperiumgames.com',
        },
        qa2: {
            apiUrl: 'https://qa-2.cloudimperiumgames.com/api/launcher/v3',
            rootUrl: 'https://qa-2.cloudimperiumgames.com',
        },
        qa3: {
            apiUrl: 'https://qa-3.cloudimperiumgames.com/api/launcher/v3',
            rootUrl: 'https://qa-3.cloudimperiumgames.com',
        },
        qa4: {
            apiUrl: 'https://qa-4.cloudimperiumgames.com/api/launcher/v3',
            rootUrl: 'https://qa-4.cloudimperiumgames.com',
        },
        qa5: {
            apiUrl: 'https://qa-5.cloudimperiumgames.com/api/launcher/v3',
            rootUrl: 'https://qa-5.cloudimperiumgames.com',
        },
        qa6: {
            apiUrl: 'https://qa-6.cloudimperiumgames.com/api/launcher/v3',
            rootUrl: 'https://qa-6.cloudimperiumgames.com',
        },
        qah3: {
            apiUrl: 'https://qa-heap3.cloudimperiumgames.com/api/launcher/v3',
            rootUrl: 'https://qa-heap3.cloudimperiumgames.com',
        },
        uat: {
            apiUrl: 'https://qa-uat.cloudimperiumgames.com/api/launcher/v3',
            rootUrl: 'https://qa-uat.cloudimperiumgames.com',
        },
        staging: {
            apiUrl: 'https://staging.cloudimperiumgames.com/api/launcher/v3',
            rootUrl: 'https://staging.cloudimperiumgames.com',
        },
        ptu: {
            apiUrl: 'https://ptu.cloudimperiumgames.com/api/launcher/v3',
            rootUrl: 'https://ptu.cloudimperiumgames.com',
        },
        eptu: {
            apiUrl: 'https://ptu.cloudimperiumgames.com/api/launcher/v3',
            rootUrl: 'https://ptu.cloudimperiumgames.com',
        },
        prod: {
            apiUrl: 'https://robertsspaceindustries.com/api/launcher/v3',
            rootUrl: 'https://robertsspaceindustries.com',
        },
    },
    infoApp: {
        appName: 'Launcher',
        companyName: 'Cloud Imperium Games',
        defaultLibraryFolder: 'Roberts Space Industries',
        defaultLibraryPath: 'C:\\Program Files\\',
    },
    sentry: {
        minidumpUrl: 'https://sentry.cloudimperiumgames.com/api/47/minidump/?sentry_key=6d91dc2ae1b74c9e80ee11bbf507df58',
        dsn: 'https://6d91dc2ae1b74c9e80ee11bbf507df58@sentry.cloudimperiumgames.com/47',
    },
    enlistUrl: 'https://robertsspaceindustries.com/enlist',
    recoverUrl: 'https://robertsspaceindustries.com/connect',
    rootUrl: 'https://robertsspaceindustries.com',
    websiteUrl: 'https://robertsspaceindustries.com',
    issueCouncilUrl: 'https://issue-council.robertsspaceindustries.com',
    copyPTUAccountSupportUrl: 'https://support.robertsspaceindustries.com/hc/en-us/articles/360000668488-Install-the-Star-Citizen-PTU',
    spectrumUrl: 'https://robertsspaceindustries.com/spectrum/community/SC',
    communityHubUrl: 'https://robertsspaceindustries.com/community-hub',
    supportUrl: 'https://support.robertsspaceindustries.com',
    statusUrl: 'https://status.robertsspaceindustries.com',
    welcomeHubUrl: 'https://robertsspaceindustries.com/playstarcitizen',
    analyticsUrl: 'https://launcher2.robertsspaceindustries.com',
    defaultAvatarUrl: 'https://cdn.robertsspaceindustries.com/static/images/account/avatar_default.jpg',
    playnowUrl: 'https://robertsspaceindustries.com/star-citizen/play-now',
    patchnotesUrl: 'https://robertsspaceindustries.com/patch-notes',
    minimumRequirementsUrl: 'https://support.robertsspaceindustries.com/hc/en-us/articles/360042417374-Star-Citizen-Minimum-System-Requirements',
    filesSupportUrl: 'https://support.robertsspaceindustries.com/hc/en-us/articles/360000065688-Send-In-Game-Files-for-RSI-Support',
    authenticationExpiry: 2592000000,
    assetsDirectory: './assets',
    libraryPollingInterval: {
        base: 300,
        window: 600,
    },
    newsPollingInterval: {
        base: 1000,
        window: 1000,
    },
    sounds: {
        open: '/sounds/phazein.wav',
        error: '/sounds/website_ui_rejection.wav',
        save: '/sounds/website_ui_savesettings.wav',
    },
    musics: {
        bg1: '/musics/SC_DL_Raven_Music_23LUFS.ogg',
        bg2: '/musics/SC_PMC_600i_v31_24bit_23LUFS.ogg',
        bg3: '/musics/SC_PMC_First Light_23LUFS.ogg',
        bg4: '/musics/SC_PMC_Main_Theme_23LUFS.ogg',
        bg5: '/musics/SC_PMC_Majesty of Space_23LUFS.ogg',
        bg6: '/musics/SC_PMC_Mind Games_23LUFS.ogg',
    },
    gameAssets: [
        {
            id: 'SC',
            logo: '/logos/sc-game-logo-small.svg',
            logoWide: '/logos/sc-game-logo-wide.svg',
            logoAnimation: '/logos/sc-logo-animation.json',
            bgImage: '/images/sc_bg_fallback.jpg',
            bgVideo: '/videos/sc_bg_video.webm',
        },
        {
            id: 'SQ42',
            logo: '/logos/sq42-game-logo-small.svg',
            bgImage: '/images/sc_bg_evocati.jpg',
        },
    ],
    carousel: {
        delay: 25000,
        images: [
            '/images/carousel-slide-1@2x.jpg',
            '/images/carousel-slide-2@2x.jpg',
            '/images/carousel-slide-3@2x.jpg',
            '/images/carousel-slide-4@2x.jpg',
        ],
    },
    loginDataRemovalTimeout: 5 * 60 * 1000,
    statusDataRefreshInterval: 60 * 1000,
    appLayoutLoaderWipeTrigger: 2500, // Force the main app loading screen to show for at least 3 seconds
};
exports.default = configuration;
