// @rsilauncher/app-shared/src/ipc-events/installer/installer.event.ts
export const installer = {
  //Fix permissions
  INSTALLER_FIX_PERMISSIONS: 'installer@fix-permissions',
  INSTALLER_FIX_PERMISSIONS_SUCCESSFUL: 'installer@fix-permissions-successful',
  INSTALLER_FIX_PERMISSIONS_FAILED: 'installer@fix-permissions-failed',

  //Start Phases
  INSTALLER_DOWNLOAD_START: 'installer@initial-download-start',
  INSTALLER_RETRIEVE_REMOTE_FILE_START: 'installer@retrieve-remote-file-list-start',
  INSTALLER_UPDATE_FILE_P4K_START: 'installer@update-files-inside-p4k-start',
  INSTALLER_UPDATE_LOOSE_FILE_START: 'installer@update-loose-files-start',
  INSTALLER_UPDATE_PK4_START: 'installer@update-p4k-structure-start',

  //Progress phases
  INSTALLER_DOWNLOAD_PROGRESS: 'installer@initial-download-progress',
  INSTALLER_RETRIEVE_REMOTE_FILE_PROGRESS: 'installer@retrieve-remote-file-list-progress',
  INSTALLER_UPDATE_FILE_P4K_PROGRESS: 'installer@update-files-inside-p4k-progress',
  INSTALLER_UPDATE_LOOSE_FILE_PROGRESS: 'installer@update-loose-files-progress',

  //Compute phases
  COMPUTE_FILE_LIST_DIFFERENCE_START: 'installer@compute-file-list-difference-start',
  COMPUTE_FILE_LIST_DIFFERENCE_END: 'installer@compute-file-list-difference-end',

  //End phases
  INSTALLER_DOWNLOAD_END: 'installer@initial-download-end',
  INSTALLER_RETRIEVE_REMOTE_FILE_END: 'installer@retrieve-remote-file-list-end',
  INSTALLER_UPDATE_FILE_P4K_END: 'installer@update-files-inside-p4k-end',
  INSTALLER_UPDATE_LOOSE_FILE_END: 'installer@update-loose-files-end',
  INSTALLER_UPDATE_PK4_END: 'installer@update-p4k-structure-end',

  //Pause
  INSTALLER_PAUSE: 'installer@pause',
  INSTALLER_PAUSE_SUCCESSFUL: 'installer@pause-successful',
  INSTALLER_PAUSE_FAILED: 'installer@pause-failed',

  //Resume
  INSTALLER_RESUME: 'installer@resume',
  INSTALLER_RESUME_SUCCESSFUL: 'installer@resume-successful',
  INSTALLER_RESUME_FAILED: 'installer@resume-failed',

  //Cancel
  INSTALLER_CANCEL: 'installer@cancel',
  INSTALLER_CANCEL_SUCCESSFUL: 'installer@cancel-successful',
  INSTALLER_CANCEL_FAILED: 'installer@cancel-failed',

  //Set options
  INSTALLER_SET_OPTIONS: 'installer@set-option',
  INSTALLER_SET_OPTION_SUCCESSFUL: 'installer@set-option-successful',

  //Install
  INSTALLER_INSTALL: 'installer@install',
  INSTALLER_INSTALL_SUCCESSFUL: 'installer@install-successful',
  INSTALLER_INSTALL_FAILED: 'installer@install-failed',
  INSTALLER_INSTALL_PROGRESS: 'installer@install-progress',
};
